packages/transport/src/chrome-runtime/client-connection-manager.ts

```diff
@@ -0,0 +1,183 @@
+   */
+  static init(label?: ChannelClientLabel.Extension): MessagePort;
+  static init(label: ChannelClientLabel.ContentScript, initPort: MessagePort): undefined;
+  static init(label = ChannelClientLabel.Extension, initPort?: MessagePort) {
+    if (label === ChannelClientLabel.Extension && !initPort) {
+      const { port1, port2 } = new MessageChannel();
+      ClientConnectionManager.singleton ??= new ClientConnectionManager(label, port1);
+      ClientConnectionManager.attached ??= port2;
+    } else if (label == ChannelClientLabel.ContentScript && initPort) {
+      ClientConnectionManager.singleton ??= new ClientConnectionManager(label, initPort);
+      ClientConnectionManager.attached ??= undefined;
+      if (ClientConnectionManager.attached) throw Error('Somehow already attached');
+    } else throw Error('Invalid arguments');
+    return ClientConnectionManager.attached;
+  }
+
+  private constructor(
+    protected label: ChannelClientLabel,
+    private initPort: MessagePort,
+  ) {
+    if (ClientConnectionManager.singleton) throw Error('Already constructed');
+    this.initPort.addEventListener('message', evt => {
+      if (isClientInitMessage(evt)) this.initConnection(evt.data);
+    });
+    this.initPort.start();
+  }
+
+  /**
+   * Requests a new connection to a service in the extension. Any untrusted page
+   * may attempt to connect.
+   *
+   * All further communication with a client happens over the page-provided
+   * clientPort.  All communication with the extension happens over the
+   * chrome.runtime.Port initialized here.
+   *
+   * @param clientPort MessagePort provided by the page-controlled client
+   * @param serviceTypeName requested fully-qualified service typename
+   */
+  private initConnection({ port: clientPort, service: serviceTypeName }: InitChannelClientData) {
+    const connection = {} as Partial<ClientConnection>;
+
+    const name = nameClientConnection(this.label, serviceTypeName);
+    const clientId = name.uuid;
+    const servicePort = chrome.runtime.connect({ includeTlsChannelId: true, name: String(name) });
+
+    /**
+     * Client-specific message handler. Listens for messages on the
+     * page-provided client channel.  Messages are forwarded to the service port
+     * without much validation.
+     *
+     * A reference is kept in case we want to quietly remove the listener later.
+     *
+     * @param ev may be any message transmitted by any untrusted page or other
+     * untrusted script with access to the page-provided message channel
+     */
+    const clientListener = (ev: MessageEvent<TransportMessage | TransportStream>) => {
+      if (!isTransportData(ev.data)) throw Error('Unknown data in transport');
+      const request = ev.data;
+      if (isTransportMessage(request)) {
+        const { requestId, message } = request;
+        servicePort.postMessage({ requestId, message });
+      } // TODO: isTransportStream(request)
```

Comment:
It looks like `isTransportStream` is available. Do we need this to serve stream requests?